// aws-cdk/test/context-providers/test.amis.js
"use strict";
const aws = require("aws-sdk");
const AWS = require("aws-sdk-mock");
const nodeunit = require("nodeunit");
const ami_1 = require("../../lib/context-providers/ami");
AWS.setSDKInstance(aws);
const mockSDK = {
    defaultAccount: () => Promise.resolve('123456789012'),
    defaultRegion: () => Promise.resolve('bermuda-triangle-1337'),
    cloudFormation: () => { throw new Error('Not Mocked'); },
    ec2: () => Promise.resolve(new aws.EC2()),
    ecr: () => { throw new Error('Not Mocked'); },
    route53: () => { throw new Error('Not Mocked'); },
    s3: () => { throw new Error('Not Mocked'); },
    ssm: () => { throw new Error('Not Mocked'); },
};
module.exports = nodeunit.testCase({
    async 'calls DescribeImages on the request'(test) {
        // GIVEN
        let request;
        AWS.mock('EC2', 'describeImages', (params, cb) => {
            request = params;
            return cb(null, { Images: [{ ImageId: 'ami-1234' }] });
        });
        // WHEN
        await new ami_1.AmiContextProviderPlugin(mockSDK).getValue({
            account: '1234',
            region: 'asdf',
            owners: ['some-owner'],
            filters: {
                'some-filter': ['filtered']
            }
        });
        // THEN
        test.deepEqual(request, {
            Owners: ['some-owner'],
            Filters: [
                {
                    Name: 'some-filter',
                    Values: ['filtered'],
                }
            ]
        });
        AWS.restore();
        test.done();
    },
    async 'returns the most recent AMI matching the criteria'(test) {
        // GIVEN
        AWS.mock('EC2', 'describeImages', (_, cb) => {
            return cb(null, { Images: [
                    {
                        ImageId: 'ami-1234',
                        CreationDate: "2016-06-22T08:39:59.000Z",
                    },
                    {
                        ImageId: 'ami-5678',
                        CreationDate: "2019-06-22T08:39:59.000Z",
                    }
                ] });
        });
        // WHEN
        const result = await new ami_1.AmiContextProviderPlugin(mockSDK).getValue({
            account: '1234',
            region: 'asdf',
            filters: {}
        });
        // THEN
        test.equals(result, 'ami-5678');
        AWS.restore();
        test.done();
    }
});
