// aws-cdk/test/context-providers/test.generic.js
"use strict";
const contextproviders = require("../../lib/context-providers");
const settings_1 = require("../../lib/settings");
const mockSDK = {
    defaultAccount: () => Promise.resolve('123456789012'),
    defaultRegion: () => Promise.resolve('bermuda-triangle-1337'),
    cloudFormation: () => { throw new Error('Not Mocked'); },
    ec2: () => { throw new Error('Not Mocked'); },
    ecr: () => { throw new Error('Not Mocked'); },
    route53: () => { throw new Error('Not Mocked'); },
    s3: () => { throw new Error('Not Mocked'); },
    ssm: () => { throw new Error('Not Mocked'); },
};
module.exports = {
    async 'errors are reported into the context value'(test) {
        // GIVEN
        contextproviders.registerContextProvider('testprovider', class {
            async getValue(_) {
                throw new Error('Something went wrong');
            }
        });
        const context = new settings_1.Context();
        // WHEN
        await contextproviders.provideContextValues([
            { key: 'asdf', props: {}, provider: 'testprovider' }
        ], context, mockSDK);
        // THEN - error is now in context
        // NOTE: error key is inlined here because it's part of the CX-API
        // compatibility surface.
        test.equals(context.get('asdf').$providerError, 'Something went wrong');
        test.done();
    },
    async 'errors are marked transient'(test) {
        // GIVEN
        contextproviders.registerContextProvider('testprovider', class {
            async getValue(_) {
                throw new Error('Something went wrong');
            }
        });
        const context = new settings_1.Context();
        // WHEN
        await contextproviders.provideContextValues([
            { key: 'asdf', props: {}, provider: 'testprovider' }
        ], context, mockSDK);
        // THEN - error is marked transient
        test.ok(context.get('asdf')[settings_1.TRANSIENT_CONTEXT_KEY]);
        test.done();
    },
};
